File: src/lib/rateLimit.ts
// Fair-use rate limiting for the hosted instance (Phase 9 Part 2, ADR 0019).
//
// Backed by Upstash Redis (`@upstash/ratelimit` sliding window). If no
// Upstash REST credentials are configured — i.e. every self-host
// deployment, and any hosted deployment before the integration is
// provisioned — this is a **no-op pass-through**. Self-host stays
// unthrottled and full-featured (ADR 0010).
//
// Env var names: the Vercel Upstash integration injects `KV_REST_API_URL` /
// `KV_REST_API_TOKEN`; a manual / self-host setup may use the Upstash-native
// `UPSTASH_REDIS_REST_URL` / `UPSTASH_REDIS_REST_TOKEN`. Either works.
//
// Key: the Clerk user id when signed in, else the client IP. Tiers carry
// different budgets; the numbers here are a starting point, tuned against
// real traffic later (they intentionally live in code, not an ADR).

import { Ratelimit } from '@upstash/ratelimit';
import { Redis } from '@upstash/redis';
import { getCurrentUserId } from '@/lib/currentUserId';

const REST_URL = process.env.KV_REST_API_URL || process.env.UPSTASH_REDIS_REST_URL;
const REST_TOKEN = process.env.KV_REST_API_TOKEN || process.env.UPSTASH_REDIS_REST_TOKEN;

/** True when a limiter store is configured. When false, everything passes. */
export const rateLimitEnabled = Boolean(REST_URL && REST_TOKEN);

export type RateLimitTier = 'default' | 'ai' | 'mcp';

interface TierConfig {
  /** requests per window for a signed-in caller */
  authed: number;
  /** requests per window for an anonymous caller (by IP) */
  anon: number;
  /** sliding window */
  window: `${number} s` | `${number} m`;
}

// Starting budgets — deliberately generous for real dashboard use, tight
// for the token-spending AI routes and for anonymous callers.
const TIERS: Record<RateLimitTier, TierConfig> = {
  default: { authed: 120, anon: 30, window: '60 s' },
  ai: { authed: 15, anon: 6, window: '60 s' },
  mcp: { authed: 120, anon: 60, window: '60 s' },
};

export interface RateLimitResult {
  ok: boolean;
  limit: number;
  remaining: number;
  /** epoch ms when the window resets */
  reset: number;
}

const PASS: RateLimitResult = { ok: true, limit: 0, remaining: 0, reset: 0 };

let redis: Redis | null = null;
const limiters = new Map<string, Ratelimit>();

function getLimiter(tier: RateLimitTier, authed: boolean): Ratelimit {
  const key = `${tier}:${authed ? 'authed' : 'anon'}`;
  const existing = limiters.get(key);
  if (existing) return existing;

  redis ??= new Redis({ url: REST_URL!, token: REST_TOKEN! });
  const cfg = TIERS[tier];
  const limit = authed ? cfg.authed : cfg.anon;
  const rl = new Ratelimit({
    redis,
    limiter: Ratelimit.slidingWindow(limit, cfg.window),
    prefix: `mm:rl:${key}`,
    analytics: false,
  });
  limiters.set(key, rl);
  return rl;
}

/** First hop of `x-forwarded-for`, or `x-real-ip`, else a fixed fallback. */
function clientIp(req: Request): string {
  const xff = req.headers.get('x-forwarded-for');
  if (xff) {
    const first = xff.split(',')[0]?.trim();
    if (first) return first;
  }
  return req.headers.get('x-real-ip')?.trim() || 'unknown-ip';
}

/**
 * Check (and consume) one unit of the caller's budget for `tier`.
 * Returns `{ ok: true, ... }` unchanged when rate limiting is disabled.
 *
 * Pass `opts.userId` when the caller already has it (e.g. middleware, which
 * gets it from Clerk's callback); otherwise it is resolved here.
 * `opts.userId === null` means "known anonymous, skip the lookup".
 */
export async function checkRateLimit(
  req: Request,
  tier: RateLimitTier = 'default',
  opts: { userId?: string | null } = {}
): Promise<RateLimitResult> {
  if (!rateLimitEnabled) return PASS;

  let userId: string | null;
  if ('userId' in opts) {
    userId = opts.userId ?? null;
  } else {
    try {
      userId = await getCurrentUserId();
    } catch {
      userId = null;
    }
  }
  // `getCurrentUserId()` returns the literal 'local' only in self-host, where
  // `rateLimitEnabled` is already false — so here a value means a real user.
  const authed = Boolean(userId && userId !== 'local');
  const key = authed ? `u:${userId}` : `ip:${clientIp(req)}`;

  try {
    const r = await getLimiter(tier, authed).limit(key);
    return { ok: r.success, limit: r.limit, remaining: r.remaining, reset: r.reset };
  } catch {
    // A limiter/store failure must not take the endpoint down — fail open.
    return PASS;
  }
}

/** Standard headers to attach to any response covered by a limiter. */
export function rateLimitHeaders(r: RateLimitResult): Record<string, string> {
  if (!r.limit) return {};
  return {
    'RateLimit-Limit': String(r.limit),
    'RateLimit-Remaining': String(Math.max(0, r.remaining)),
    'RateLimit-Reset': String(Math.max(0, Math.ceil((r.reset - Date.now()) / 1000))),
  };
}

/** The 429 body + headers for a rejected request. */
export function rateLimitResponse(r: RateLimitResult): Response {
  const retryAfter = Math.max(1, Math.ceil((r.reset - Date.now()) / 1000));
  return new Response(
    JSON.stringify({
      success: false,
      data: null,
      error: 'Rate limit exceeded. Slow down and try again shortly.',
    }),
    {
      status: 429,
      headers: {
        'Content-Type': 'application/json',
        'Retry-After': String(retryAfter),
        ...rateLimitHeaders(r),
      },
    }
  );
}

type RouteHandler = (req: Request, ctx?: unknown) => Promise<Response> | Response;

/**
 * Wrap a route handler so every call first passes `tier`'s budget. On a
 * rejection it returns the 429 without invoking the handler; otherwise it
 * runs the handler and adds `RateLimit-*` headers to its response.
 */
export function withRateLimit(handler: RouteHandler, tier: RateLimitTier = 'default'): RouteHandler {
  return async (req, ctx) => {
    const r = await checkRateLimit(req, tier);
    if (!r.ok) return rateLimitResponse(r);
    const res = await handler(req, ctx);
    const extra = rateLimitHeaders(r);
    for (const [k, v] of Object.entries(extra)) res.headers.set(k, v);
    return res;
  };
}

File: services/fundamentals-api/app/rate_limit.py
"""Fair-use rate limiting for the public fundamentals API (Phase 9, ADR 0019).

Fixed-window counter per client IP, stored in Upstash Redis via its REST API
(`httpx` only — no `redis` dependency, works fine in a Vercel serverless
function). If the Upstash env vars are unset the limiter is a **no-op
pass-through**, so local dev and any self-host deployment are never
throttled.

The main Next.js app already rate-limits `/api/*` and the MCP endpoint; this
closes the one remaining path — hitting this service's public URL directly.
"""

from __future__ import annotations

import logging
import time
from dataclasses import dataclass

import httpx

from app.config import Settings

logger = logging.getLogger(__name__)

_WINDOW_SECONDS = 60
# A little longer than the window so a key that's still being read against
# can't be evicted mid-window.
_KEY_TTL_SECONDS = 65


@dataclass(frozen=True)
class RateLimitResult:
    allowed: bool
    limit: int
    remaining: int
    retry_after: int  # seconds until the window rolls over


def _passthrough(settings: Settings) -> RateLimitResult:
    return RateLimitResult(True, settings.rate_limit_per_minute, settings.rate_limit_per_minute, 0)


def is_enabled(settings: Settings) -> bool:
    return bool(settings.upstash_redis_rest_url and settings.upstash_redis_rest_token)


async def check_rate_limit(ip: str, settings: Settings) -> RateLimitResult:
    """Consume one unit of `ip`'s budget for the current minute-window.

    Fails **open**: any transport/store error returns `allowed=True` rather
    than taking the endpoint down.
    """
    if not is_enabled(settings):
        return _passthrough(settings)

    now = int(time.time())
    window = now // _WINDOW_SECONDS
    key = f"mmf:rl:{ip}:{window}"
    reset_in = _WINDOW_SECONDS - (now % _WINDOW_SECONDS)

    try:
        async with httpx.AsyncClient(timeout=2.0) as client:
            resp = await client.post(
                f"{settings.upstash_redis_rest_url.rstrip('/')}/pipeline",
                headers={"Authorization": f"Bearer {settings.upstash_redis_rest_token}"},
                json=[["INCR", key], ["EXPIRE", key, str(_KEY_TTL_SECONDS), "NX"]],
            )
        resp.raise_for_status()
        count = int(resp.json()[0]["result"])
    except Exception as exc:  # noqa: BLE001 — deliberate fail-open
        logger.warning("rate-limit check failed, allowing request: %s", exc)
        return _passthrough(settings)

    limit = settings.rate_limit_per_minute
    remaining = max(0, limit - count)
    return RateLimitResult(count <= limit, limit, remaining, reset_in)


def client_ip(headers: httpx.Headers | dict, fallback: str | None) -> str:
    """First hop of `x-forwarded-for` (Vercel sets it), else the socket peer."""
    xff = headers.get("x-forwarded-for") if hasattr(headers, "get") else None
    if xff:
        first = xff.split(",")[0].strip()
        if first:
            return first
    return fallback or "unknown-ip"
